src/quarot/hadamard_utils.py

```diff
@@ -0,0 +1,143 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import math
+
+import torch
+from fast_hadamard_transform import hadamard_transform
+
+from .hadamard_tensors import (
+    get_had12,
+    get_had20,
+    get_had28,
+    get_had36,
+    get_had40,
+    get_had52,
+    get_had60,
+    get_had108,
+    get_had140,
+    get_had156,
+    get_had172,
+)
+
+
+def get_hadK(n: int) -> tuple[torch.Tensor, int]:
+    if is_pow2(n):
+        return None, 1
+
+    Ks = [172, 156, 140, 108, 60, 52, 40, 36, 28, 20, 12]
+    for K in Ks:
+        if n % K == 0:
+            assert is_pow2(n // K)
+            return get_hadamard_tensor(K), K
+
+    raise ValueError(f"Unsupported Hadamard dimension: {n}")
+
+
+def get_hadamard_tensor(dim: int) -> torch.Tensor:
+    hadamard_functions = {
+        172: get_had172,
+        156: get_had156,
+        140: get_had140,
+        108: get_had108,
+        60: get_had60,
+        52: get_had52,
+        40: get_had40,
+        36: get_had36,
+        28: get_had28,
+        20: get_had20,
+        12: get_had12,
+    }
+    return hadamard_functions[dim]()
+
+
+def matmul_hadU(X: torch.Tensor) -> torch.Tensor:
```

Comment:
@sashkboos is this method meant to be exactly the same as [the cuda one](https://github.com/microsoft/TransformerCompression/blob/49c544b537e6a73ff11322e93ba348a59ccc0282/src/quarot/hadamard_utils.py#L92) (later in this file), but just pytorch instead of cuda version?

Reply:
@nailimixaM 

Yes. but this is a more general version of that. Basically, this applies Hadamard on arbitrary dimension which is slow (even on GPU)

Reply:
refactored to make all methods in this file clearer.

Reply:
LGTM. Thanks Max :)